### PR TITLE
Fix bug that whether an Orchestrator object can extract entities or n…

### LINKED ISSUE #5750: Whether a static Orchestrator object can extract entities or not should not depend on a member variable, ScoreEntities
## Version
What package version of the SDK are you using.

## Describe the bug
Give a clear and concise description of what the bug is.

## Expected behavior
Give a clear and concise description of what you expected to happen.

### PR DESCRIPTION
**Pull request by hcyang**
…ot should not depend on a member variable ScoreEntities.

Fixes #5750 

## Specific Changes

  - 
  -
  -

### CODE REVIEW

**Review by tsuwandy — APPROVED** [review]

**Comment by tsuwandy** [comment]
@hcyang, reviewed the changes again and not sure why we need the extra layer, OrchestratorDictionaryEntry?

**Comment by hcyang** [comment]
> @hcyang, @daveta and I reviewed the changes again and not sure why we need the extra layer, OrchestratorDictionaryEntry?

@tsuwandy, the removed ScoreEntities flag is renamed IsEntityReady and placed in an OrchestratorDictionaryEntry and associated with a static Orchestrator object. If there is no need for the ScoreEntities flag, then we can remove that IsEntityReady flag and remove the whole OrchestratorDictionaryEntry abstraction as well.

**Comment by daveta** [comment]
Seems like the current code and this fix are equivalent (?).  The difference is the new code, you have available the Orchestrator more readily available in the Recognizer instance (but I don't see it being used).

File: libraries/Microsoft.Bot.Builder.AI.Orchestrator/OrchestratorRecognizer.cs
Comment (carlosscastro) [inline libraries/Microsoft.Bot.Builder.AI.Orchestrator/OrchestratorRecognizer.cs:343]: I know this line was moved from above, but in any case it seems like a potentially dangerous one. If we just instantiated a JArray, we are safe (i.e. if the `if` statement in line 338 was executed). However, if the `if statement` did not get executed, we got a `JToken` from the json payload. Given that `JToken` is more general than `JArray`, this cast could lead to an invalid cast. Do we guarantee that 100% of the time this will be a JArray? If yes, then why isn't the `values` variable in line 337 declared as a JArray instead? to handle empty entries? If this is all on purpose, some comments explaining the decision making behind this could be useful for future maintainers

File: libraries/Microsoft.Bot.Builder.AI.Orchestrator/OrchestratorRecognizer.cs
Comment (carlosscastro) [inline libraries/Microsoft.Bot.Builder.AI.Orchestrator/OrchestratorRecognizer.cs:323]: Since we are visiting this codebase and adding some restructuring, it would be nice, if possible, to add some comments and explain what is the json format, what we are trying to extract, and other decisions that are not obvious in the code.

File: libraries/Microsoft.Bot.Builder.AI.Orchestrator/OrchestratorRecognizer.cs
Comment (carlosscastro) [inline libraries/Microsoft.Bot.Builder.AI.Orchestrator/OrchestratorRecognizer.cs:361]: Same comment as above, are we 100% this is a JArray? If yes could you help me understand for future reference, why it is instanceData declared as JToken?

File: libraries/Microsoft.Bot.Builder.AI.Orchestrator/OrchestratorRecognizer.cs
Comment (carlosscastro) [inline libraries/Microsoft.Bot.Builder.AI.Orchestrator/OrchestratorRecognizer.cs]: no real need to initialize as null.

File: libraries/Microsoft.Bot.Builder.AI.Orchestrator/OrchestratorRecognizer.cs
Comment (carlosscastro) [inline libraries/Microsoft.Bot.Builder.AI.Orchestrator/OrchestratorRecognizer.cs:347]: $instance is used in many places, consider adding a constant so we centralize the definition.

File: libraries/Microsoft.Bot.Builder.AI.Orchestrator/OrchestratorRecognizer.cs
Comment (carlosscastro) [inline libraries/Microsoft.Bot.Builder.AI.Orchestrator/OrchestratorRecognizer.cs]: Just trying to learn about this piece of code: I assume this is executed only once on load, and not on a per-turn basis right?

Comment (carlosscastro) [reply libraries/Microsoft.Bot.Builder.AI.Orchestrator/OrchestratorRecognizer.cs]: Also, lines 424 and 425 read the file twice and use twice as much memory as needed. Can both be replaced by 

```
byte[] snapshotBytes=  File.ReadAllBytes(fullSnapShotFile);
```

Avoiding the binary -> string -> binary extra copying and memory

File: libraries/Microsoft.Bot.Builder.AI.Orchestrator/OrchestratorRecognizer.cs
Comment (carlosscastro) [inline libraries/Microsoft.Bot.Builder.AI.Orchestrator/OrchestratorRecognizer.cs]: Is this class used outside of this assembly? If not, please make it private. Protected members get exported publicly and we need to maintain backward compat for the foreseeable future.

File: libraries/Microsoft.Bot.Builder.AI.Orchestrator/OrchestratorRecognizer.cs
Comment (carlosscastro) [inline libraries/Microsoft.Bot.Builder.AI.Orchestrator/OrchestratorRecognizer.cs:85]: I see that this property shipped with R13 at least. We cannot delete it without breaking backward compatibility.

File: libraries/Microsoft.Bot.Builder.AI.Orchestrator/OrchestratorRecognizer.cs
Comment (carlosscastro) [inline libraries/Microsoft.Bot.Builder.AI.Orchestrator/OrchestratorRecognizer.cs]: So in this method we check for null, and if _resolver is null, we do heavy I/O and then set the resolver. CAn you guarantee that only one instance of this class is going to have the initialize model executed simultaneously? I assume not. If not, you would have a race condition here, where in the best case scenario, multiple threads do all the processing in parallel and would all be setting the same variable. Adding lazy initialization or some kind of protection for race conditions would remove that risk. Happy to chat about lifecycle / access patterns of this class to understand what the actual risks are, since I'm just observing in vacuum without as much context.

File: libraries/Microsoft.Bot.Builder.AI.Orchestrator/OrchestratorRecognizer.cs
Comment (carlosscastro) [reply libraries/Microsoft.Bot.Builder.AI.Orchestrator/OrchestratorRecognizer.cs]: It would be tricky, but if I inherit from OrchestratorRecognizer as a bot developer, I'd be able to access this class and us such, this becomes part of our public surface and needs to be treated as such. From what I observe, there is no value in making that commitment at this point. Could we mark it as private class instead?

Comment (carlosscastro) [reply libraries/Microsoft.Bot.Builder.AI.Orchestrator/OrchestratorRecognizer.cs]: Example:

Assembly A:

public class Outer { protected class Inner {}}

Assembly B (references A):

public class Foo {  public Foo() { var inner = new Inner();}} // this works!

**Review by carlosscastro — CHANGES_REQUESTED** [review]
Hello! Some of the general comments I added are me just catching up and asking questions about this codebase and not issues with this PR in particular. I did find 2 things that we should fix before merging as I was learning by reading it, which I wanted to call out:

[1- We cannot remove public properties that shipped](https://github.com/microsoft/botbuilder-dotnet/pull/5749/files#diff-af81a5d0d91e35cbfcc21951b2eaf4e5ee5287aba8cb748e2be222456bd1a309L86)
2- We should minimize visibility of classes wherever possible. I'm not sure that OrchestratorDictionaryEntry needs to be protected, since as such we'd need to maintain backward compat until V5 at least

Happy to help with these, and we can chat about the other comments offline and not block this PR on them.

**Comment by carlosscastro** [comment]
Hung-chih and I chatted. The Entry class is going away, and we're discussing the backward compat issue. The rest of the comments from pre-existing code can be addressed / analyzed outside of the scope of this PR.

File: libraries/Microsoft.Bot.Builder.AI.Orchestrator/OrchestratorRecognizer.cs
Comment (hcyang) [reply libraries/Microsoft.Bot.Builder.AI.Orchestrator/OrchestratorRecognizer.cs:85]: Will put it back.

File: libraries/Microsoft.Bot.Builder.AI.Orchestrator/OrchestratorRecognizer.cs
Comment (tsuwandy) [reply libraries/Microsoft.Bot.Builder.AI.Orchestrator/OrchestratorRecognizer.cs:323]: @carlosscastro, it's modeled after what's in LuisRecognizer.

File: libraries/Microsoft.Bot.Builder.AI.Orchestrator/OrchestratorRecognizer.cs
Comment (tsuwandy) [reply libraries/Microsoft.Bot.Builder.AI.Orchestrator/OrchestratorRecognizer.cs:361]: @carlosscastro, following this previous precedent: https://github.com/microsoft/botbuilder-dotnet/blob/402bc02b4cbbd2f4ec359134640e99211367e4a5/libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/Recognizers/EntityRecognizers/EntityRecognizer.cs

File: libraries/Microsoft.Bot.Builder.AI.Orchestrator/OrchestratorRecognizer.cs
Comment (hcyang) [reply libraries/Microsoft.Bot.Builder.AI.Orchestrator/OrchestratorRecognizer.cs]: Agree, will use ReadAllBytes()

File: libraries/Microsoft.Bot.Builder.AI.Orchestrator/OrchestratorRecognizer.cs
Comment (hcyang) [reply libraries/Microsoft.Bot.Builder.AI.Orchestrator/OrchestratorRecognizer.cs]: I know, just prefer to be explicit.

File: libraries/Microsoft.Bot.Builder.AI.Orchestrator/OrchestratorRecognizer.cs
Comment (tsuwandy) [reply libraries/Microsoft.Bot.Builder.AI.Orchestrator/OrchestratorRecognizer.cs]: @carlosscastro, yes, only one instance is going to initialize the model, code is run with ConcurrentDictionary.GetOrAdd that's thread safe.

File: libraries/Microsoft.Bot.Builder.AI.Orchestrator/OrchestratorRecognizer.cs
Comment (hcyang) [reply libraries/Microsoft.Bot.Builder.AI.Orchestrator/OrchestratorRecognizer.cs]: And yes, the model file was loaded only once per load, not per turn.

File: libraries/Microsoft.Bot.Builder.AI.Orchestrator/OrchestratorRecognizer.cs
Comment (hcyang) [reply libraries/Microsoft.Bot.Builder.AI.Orchestrator/OrchestratorRecognizer.cs]: Will make the class private.

File: libraries/Microsoft.Bot.Builder.AI.Orchestrator/OrchestratorRecognizer.cs
Comment (hcyang) [reply libraries/Microsoft.Bot.Builder.AI.Orchestrator/OrchestratorRecognizer.cs]: It's possible that an OrchestratorRecognizer object is constructed using the default constructor with necessary properties set. The object's RecognizeAsync() method is then called by a pool of threads. In such scenario, the InitializeModel() function can become a race condition for creating the _resolver object, albeit the Orchestrator object creation is thread safe behind the ConcurrentDictionary container.
One solution is to move InitializeModel() call out of RecognizeAsync(), but it may break the "public" API contract.
Another solution is to put all race condition inside InitializeModel() and make it thread safe. Will implement the second option.

**Comment by hcyang** [comment]
> Hung-chih and I chatted. The Entry class is going away, and we're discussing the backward compat issue. The rest of the comments from pre-existing code can be addressed / analyzed outside of the scope of this PR.

@carlosscastro, after consulting with @daveta, we decided to bring the DictionaryEntry class back, but make it a private class per your suggeston.

File: libraries/Microsoft.Bot.Builder.AI.Orchestrator/OrchestratorRecognizer.cs
Comment (hcyang) [reply libraries/Microsoft.Bot.Builder.AI.Orchestrator/OrchestratorRecognizer.cs:347]: Not sure what $instance is for. Will not touch it in this PR.

File: libraries/Microsoft.Bot.Builder.AI.Orchestrator/OrchestratorRecognizer.cs
Comment (hcyang) [reply libraries/Microsoft.Bot.Builder.AI.Orchestrator/OrchestratorRecognizer.cs:363]: Will make the declaration to JArray.

File: libraries/Microsoft.Bot.Builder.AI.Orchestrator/OrchestratorRecognizer.cs
Comment (carlosscastro) [inline libraries/Microsoft.Bot.Builder.AI.Orchestrator/OrchestratorRecognizer.cs:67]: Is this constructor only for tests? Asking because of the mockup variable name

File: libraries/Microsoft.Bot.Builder.AI.Orchestrator/OrchestratorRecognizer.cs
Comment (carlosscastro) [inline libraries/Microsoft.Bot.Builder.AI.Orchestrator/OrchestratorRecognizer.cs:415]: What's the latency of this method on moderate sized models?

File: libraries/Microsoft.Bot.Builder.AI.Orchestrator/OrchestratorRecognizer.cs
Comment (hcyang) [reply libraries/Microsoft.Bot.Builder.AI.Orchestrator/OrchestratorRecognizer.cs]: I was about to renamed it to resolverExternal.

Comment (carlosscastro) [reply libraries/Microsoft.Bot.Builder.AI.Orchestrator/OrchestratorRecognizer.cs]: Given that this shipped as public in our surface area, a name without the mockup word will be better, i.e. resolver. And then you cna rename `_isResolverMockup` to `_isResolverInjected` if this variable is still needed. Mostly because since it is public, developers could pass in a legitimate label resolver and not a mockup so the name would be misleading

File: libraries/Microsoft.Bot.Builder.AI.Orchestrator/OrchestratorRecognizer.cs
Comment (carlosscastro) [reply libraries/Microsoft.Bot.Builder.AI.Orchestrator/OrchestratorRecognizer.cs:399]: The fact that the same code is elsewhere does not make it correct unfortunately. However we can investigate this separately from the current PR, so let's table this one! Thanks for the pointer.

File: libraries/Microsoft.Bot.Builder.AI.Orchestrator/OrchestratorRecognizer.cs
Comment (hcyang) [reply libraries/Microsoft.Bot.Builder.AI.Orchestrator/OrchestratorRecognizer.cs:433]: The latency for that is just seconds. Loading the models are quick even though they can be up to 1GB. The actual time consuming part is building the resolver consuming some big snapshot file. Perhaps that's the reason the InitializeModel() function is called once inside RecognizeAsync(), which is async.

Comment (carlosscastro) [reply libraries/Microsoft.Bot.Builder.AI.Orchestrator/OrchestratorRecognizer.cs:433]: what is the latency of the entire initializeModel?

Comment (carlosscastro) [reply libraries/Microsoft.Bot.Builder.AI.Orchestrator/OrchestratorRecognizer.cs:433]: Asking because 1) this will get run on the first turn and not when the app service starts. This means that the first turn of the first conversation will be really slow, i.e. as long as this method takes.

Comment (carlosscastro) [reply libraries/Microsoft.Bot.Builder.AI.Orchestrator/OrchestratorRecognizer.cs:433]: Ideally we'd want to transfer this time to application load somehow in the future. We don't need to resolve this now, but it is importnat. I'll create an issue for this.

**Review by carlosscastro — APPROVED** [review]
Thanks for the iteration @hcyang !!!

**Pull request merged**
